check whether GTID is `ON` in the upstream ealier

## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

2. What did you expect to see?

3. What did you see instead?

    ```
    [2020/08/27 16:07:03.684 +08:00] [INFO] [checkpoint.go:798] ["parsing metadata from file"] [task=nmm3] [unit="binlog replication"] [component="remote checkpoint"] [file=dumped_data.nmm3/metadata]
    [2020/08/27 16:07:03.684 +08:00] [INFO] [checkpoint.go:761] ["loaded checkpoints from meta"] [task=nmm3] [unit="binlog replication"] [component="remote checkpoint"] ["global checkpoint"="position: (mysql-bin.001974, 443173257), gtid-set: (flushed position: (mysql-bin.001974, 443173257), gtid-set: )"]
    [2020/08/27 16:07:03.880 +08:00] [INFO] [syncer.go:885] ["flushed checkpoint"] [task=nmm3] [unit="binlog replication"] [checkpoint="position: (mysql-bin.001974, 443173257), gtid-set: (flushed position: (mysql-bin.001974, 443173257), gtid-set: )"]
    [2020/08/27 16:07:03.880 +08:00] [INFO] [syncer.go:1098] ["try to remove loaded files"] [task=nmm3] [unit="binlog replication"]

    [2020/08/27 16:07:03.887 +08:00] [ERROR] [syncer.go:1290] ["fail to fetch binlog"] [task=nmm3] [unit="binlog replication"] [error="ERROR 1236 (HY000): The replication sender thread cannot start in AUTO_POSITION mode: this server has GTID_MODE = OFF instead of ON."]
   ...
    [2020/08/27 16:07:03.889 +08:00] [ERROR] [subtask.go:295] ["unit process error"] [subtask=nmm3] [unit=Sync] ["error information"="{\"ErrCode\":50000,\"ErrClass\":\"not-set\",\"ErrScope\":\"not-set\",\"ErrLevel\":\"high\",\"Message\":\"ERROR 1236 (HY000): The replication sender thread cannot start in AUTO_POSITION mode: this server has GTID_MODE = OFF instead of ON.\"}"]
    [2020/08/27 16:07:06.586 +08:00] [INFO] [worker.go:288] ["auto_resume sub task"] [component="worker controller"] [task=nmm3]
    ...
    [2020/08/27 16:07:06.634 +08:00] [INFO] [checkpoint.go:798] ["parsing metadata from file"] [task=nmm3] [unit="binlog replication"] [component="remote checkpoint"] [file=dumped_data.nmm3/metadata]
    [2020/08/27 16:07:06.635 +08:00] [ERROR] [subtask.go:295] ["unit process error"] [subtask=nmm3] [unit=Sync] ["error information"="{\"ErrCode\":11001,\"ErrClass\":\"functional\",\"ErrScope\":\"internal\",\"ErrLevel\":\"high\",\"Message\":\"parse mydumper metadata error: open dumped_data.nmm3/metadata: no such file or directory\"}"]
    ```

    `query-status` report `parse mydumper metadata error: open dumped_data.nmm3/metadata: no such file or directory`, but the real cause is `ERROR 1236 (HY000): The replication sender thread cannot start in AUTO_POSITION mode: this server has GTID_MODE = OFF instead of ON.`

    [more.log](https://github.com/pingcap/dm/files/5135281/more.log)

    Can also see it at [AskTUG](https://asktug.com/t/topic/36828).

   we can check whether `GTID_MODE=on` when `operate-source create`.

   

4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
        v2.0.0-rc
        ```

--- a/_utils/terror_gen/errors_release.txt
+++ b/_utils/terror_gen/errors_release.txt
@@ -173,6 +173,7 @@ ErrTaskCheckGenTableRouter,[code=26003:class=task-check:scope=internal:level=med
 ErrTaskCheckGenColumnMapping,[code=26004:class=task-check:scope=internal:level=medium], "Message: generate column mapping error, Workaround: Please check the `column-mappings` config in task configuration file."
 ErrTaskCheckSyncConfigError,[code=26005:class=task-check:scope=internal:level=medium], "Message: %s: %v\n detail: %v"
 ErrTaskCheckGenBAList,[code=26006:class=task-check:scope=internal:level=medium], "Message: generate block allow list error, Workaround: Please check the `block-allow-list` config in task configuration file."
+ErrSourceCheckGTID,[code=26007:class=task-check:scope=internal:level=medium], "Message: %s has GTID_MODE = %s instead of ON, Workaround: Please check the `enable-gtid` config in source configuration file."
 ErrRelayParseUUIDIndex,[code=28001:class=relay-event-lib:scope=internal:level=high], "Message: parse server-uuid.index"
 ErrRelayParseUUIDSuffix,[code=28002:class=relay-event-lib:scope=internal:level=high], "Message: UUID (with suffix) %s not valid"
 ErrRelayUUIDWithSuffixNotFound,[code=28003:class=relay-event-lib:scope=internal:level=high], "Message: no UUID (with suffix) matched %s found in %s, all UUIDs are %v"

--- a/dm/config/source_config.go
+++ b/dm/config/source_config.go
@@ -241,6 +241,16 @@ func (c *SourceConfig) Adjust(db *sql.DB) (err error) {
 		}
 	}
 
+	if c.EnableGTID {
+		val, err := utils.GetGTID(db)
+		if err != nil {
+			return err
+		}
+		if val != "ON" {
+			return terror.ErrSourceCheckGTID.Generate(c.SourceID, val)
+		}
+	}
+
 	return nil
 }
 
@@ -316,20 +326,5 @@ func (c *SourceConfig) check(metaData *toml.MetaData, err error) error {
 		return terror.ErrWorkerUndecodedItemFromFile.Generate(strings.Join(undecodedItems, ","))
 	}
 	c.adjust()
-	return nil
-}
-
-// PreCheck check valify source config
-func (c *SourceConfig) PreCheck(db *sql.DB) error {
-	if c.EnableGTID {
-		val, err := utils.GetGTID(db)
-		if err != nil {
-			return err
-		}
-		if val != "ON" {
-			return terror.ErrSourceCheckGTID.Generate(c.SourceID, val)
-		}
-	}
-
 	return nil
 }

--- a/dm/config/source_config.go
+++ b/dm/config/source_config.go
@@ -318,3 +318,18 @@ func (c *SourceConfig) check(metaData *toml.MetaData, err error) error {
 	c.adjust()
 	return nil
 }
+
+// PreCheck check valify source config
+func (c *SourceConfig) PreCheck(db *sql.DB) error {
+	if c.EnableGTID {
+		val, err := utils.GetGTID(db)
+		if err != nil {
+			return err
+		}
+		if val != "ON" {
+			return terror.ErrSourceCheckGTID.Generate(c.SourceID, val)
+		}
+	}
+
+	return nil
+}

--- a/dm/master/server.go
+++ b/dm/master/server.go
@@ -1351,9 +1351,6 @@ func parseAndAdjustSourceConfig(contents []string) ([]*config.SourceConfig, erro
 		if _, err = cfg.Yaml(); err != nil {
 			return cfgs, err
 		}
-		if err = cfg.PreCheck(fromDB.DB); err != nil {
-			return cfgs, err
-		}
 
 		cfgs[i] = cfg
 	}

--- a/dm/master/server.go
+++ b/dm/master/server.go
@@ -1351,6 +1351,10 @@ func parseAndAdjustSourceConfig(contents []string) ([]*config.SourceConfig, erro
 		if _, err = cfg.Yaml(); err != nil {
 			return cfgs, err
 		}
+		if err = cfg.PreCheck(fromDB.DB); err != nil {
+			return cfgs, err
+		}
+
 		cfgs[i] = cfg
 	}
 	return cfgs, nil

--- a/errors.toml
+++ b/errors.toml
@@ -1048,6 +1048,12 @@ description = ""
 workaround = "Please check the `block-allow-list` config in task configuration file."
 tags = ["internal", "medium"]
 
+[error.DM-task-check-26007]
+message = "%s has GTID_MODE = %s instead of ON"
+description = ""
+workaround = "Please check the `enable-gtid` config in source configuration file."
+tags = ["internal", "medium"]
+
 [error.DM-relay-event-lib-28001]
 message = "parse server-uuid.index"
 description = ""

--- a/pkg/terror/error_list.go
+++ b/pkg/terror/error_list.go
@@ -238,6 +238,7 @@ const (
 	codeTaskCheckGenColumnMapping
 	codeTaskCheckSyncConfigError
 	codeTaskCheckGenBAList
+	codeSourceCheckGTID
 )
 
 // Relay log utils error code
@@ -811,6 +812,7 @@ var (
 	ErrTaskCheckGenColumnMapping = New(codeTaskCheckGenColumnMapping, ClassTaskCheck, ScopeInternal, LevelMedium, "generate column mapping error", "Please check the `column-mappings` config in task configuration file.")
 	ErrTaskCheckSyncConfigError  = New(codeTaskCheckSyncConfigError, ClassTaskCheck, ScopeInternal, LevelMedium, "%s: %v\n detail: %v", "")
 	ErrTaskCheckGenBAList        = New(codeTaskCheckGenBAList, ClassTaskCheck, ScopeInternal, LevelMedium, "generate block allow list error", "Please check the `block-allow-list` config in task configuration file.")
+	ErrSourceCheckGTID           = New(codeSourceCheckGTID, ClassTaskCheck, ScopeInternal, LevelMedium, "%s has GTID_MODE = %s instead of ON", "Please check the `enable-gtid` config in source configuration file.")
 
 	// Relay log basic API error
 	ErrRelayParseUUIDIndex         = New(codeRelayParseUUIDIndex, ClassRelayEventLib, ScopeInternal, LevelHigh, "parse server-uuid.index", "")

--- a/pkg/terror/terror.go
+++ b/pkg/terror/terror.go
@@ -38,6 +38,7 @@ const (
 	ClassBinlogOp
 	ClassCheckpoint
 	ClassTaskCheck
+	ClassSourceCheck
 	ClassRelayEventLib
 	ClassRelayUnit
 	ClassDumpUnit
@@ -59,6 +60,7 @@ var errClass2Str = map[ErrClass]string{
 	ClassBinlogOp:      "binlog-op",
 	ClassCheckpoint:    "checkpoint",
 	ClassTaskCheck:     "task-check",
+	ClassSourceCheck:   "source-check",
 	ClassRelayEventLib: "relay-event-lib",
 	ClassRelayUnit:     "relay-unit",
 	ClassDumpUnit:      "dump-unit",

--- a/pkg/utils/db.go
+++ b/pkg/utils/db.go
@@ -403,3 +403,9 @@ func IsErrBinlogPurged(err error) bool {
 func IsNoSuchThreadError(err error) bool {
 	return IsMySQLError(err, tmysql.ErrNoSuchThread)
 }
+
+// GetGTID return GTID_MODE
+func GetGTID(db *sql.DB) (string, error) {
+	val, err := GetGlobalVariable(db, "GTID_MODE")
+	return val, err
+}

